File: android/app/src/main/java/com/rork/calzyandroid/data/I18n.kt
package com.rork.calzyandroid.data

import android.content.Context
import java.util.Locale
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/** A language the app can present itself in. Mirrors `web/src/lib/i18n.ts`. */
data class AppLanguage(
    val code: String,
    /** Name in English, so searching "Spanish" finds Español. */
    val englishName: String,
    /** Endonym — what speakers call the language themselves. */
    val nativeName: String,
    val flag: String,
    val isRTL: Boolean = false,
)

object I18n {

    /** Ordered roughly by global reach, English first. */
    val languages: List<AppLanguage> = listOf(
        AppLanguage("en", "English", "English", "\uD83C\uDDFA\uD83C\uDDF8"),
        AppLanguage("zh-Hans", "Chinese (Simplified)", "简体中文", "\uD83C\uDDE8\uD83C\uDDF3"),
        AppLanguage("zh-Hant", "Chinese (Traditional)", "繁體中文", "\uD83C\uDDF9\uD83C\uDDFC"),
        AppLanguage("es", "Spanish", "Español", "\uD83C\uDDEA\uD83C\uDDF8"),
        AppLanguage("hi", "Hindi", "हिन्दी", "\uD83C\uDDEE\uD83C\uDDF3"),
        AppLanguage("ar", "Arabic", "العربية", "\uD83C\uDDF8\uD83C\uDDE6", isRTL = true),
        AppLanguage("pt", "Portuguese", "Português", "\uD83C\uDDE7\uD83C\uDDF7"),
        AppLanguage("bn", "Bengali", "বাংলা", "\uD83C\uDDE7\uD83C\uDDE9"),
        AppLanguage("ru", "Russian", "Русский", "\uD83C\uDDF7\uD83C\uDDFA"),
        AppLanguage("ja", "Japanese", "日本語", "\uD83C\uDDEF\uD83C\uDDF5"),
        AppLanguage("de", "German", "Deutsch", "\uD83C\uDDE9\uD83C\uDDEA"),
        AppLanguage("fr", "French", "Français", "\uD83C\uDDEB\uD83C\uDDF7"),
        AppLanguage("ko", "Korean", "한국어", "\uD83C\uDDF0\uD83C\uDDF7"),
        AppLanguage("it", "Italian", "Italiano", "\uD83C\uDDEE\uD83C\uDDF9"),
        AppLanguage("tr", "Turkish", "Türkçe", "\uD83C\uDDF9\uD83C\uDDF7"),
        AppLanguage("vi", "Vietnamese", "Tiếng Việt", "\uD83C\uDDFB\uD83C\uDDF3"),
        AppLanguage("id", "Indonesian", "Bahasa Indonesia", "\uD83C\uDDEE\uD83C\uDDE9"),
        AppLanguage("ms", "Malay", "Bahasa Melayu", "\uD83C\uDDF2\uD83C\uDDFE"),
        AppLanguage("th", "Thai", "ไทย", "\uD83C\uDDF9\uD83C\uDDED"),
        AppLanguage("pl", "Polish", "Polski", "\uD83C\uDDF5\uD83C\uDDF1"),
        AppLanguage("nl", "Dutch", "Nederlands", "\uD83C\uDDF3\uD83C\uDDF1"),
        AppLanguage("uk", "Ukrainian", "Українська", "\uD83C\uDDFA\uD83C\uDDE6"),
        AppLanguage("fa", "Persian", "فارسی", "\uD83C\uDDEE\uD83C\uDDF7", isRTL = true),
        AppLanguage("he", "Hebrew", "עברית", "\uD83C\uDDEE\uD83C\uDDF1", isRTL = true),
        AppLanguage("el", "Greek", "Ελληνικά", "\uD83C\uDDEC\uD83C\uDDF7"),
        AppLanguage("cs", "Czech", "Čeština", "\uD83C\uDDE8\uD83C\uDDFF"),
        AppLanguage("ro", "Romanian", "Română", "\uD83C\uDDF7\uD83C\uDDF4"),
        AppLanguage("hu", "Hungarian", "Magyar", "\uD83C\uDDED\uD83C\uDDFA"),
        AppLanguage("sv", "Swedish", "Svenska", "\uD83C\uDDF8\uD83C\uDDEA"),
        AppLanguage("nb", "Norwegian", "Norsk", "\uD83C\uDDF3\uD83C\uDDF4"),
        AppLanguage("da", "Danish", "Dansk", "\uD83C\uDDE9\uD83C\uDDF0"),
        AppLanguage("fi", "Finnish", "Suomi", "\uD83C\uDDEB\uD83C\uDDEE"),
    )

    private val english: AppLanguage = languages.first()

    private var tables: Map<String, Map<String, String>> = emptyMap()

    /** Loads the shared 32-language string catalogue from assets/strings.json. */
    fun load(context: Context) {
        if (tables.isNotEmpty()) return
        tables = try {
            val raw = context.assets.open("strings.json").bufferedReader().use { it.readText() }
            val root = Json.parseToJsonElement(raw).jsonObject
            root.entries.associate { (code, table) ->
                code to (table as JsonObject).entries.associate { (key, value) ->
                    key to value.jsonPrimitive.content
                }
            }
        } catch (error: Exception) {
            emptyMap()
        }
    }

    fun languageFor(code: String?): AppLanguage? =
        code?.let { c -> languages.find { it.code == c } }

    /** Best match for the device locale, falling back to English. */
    fun systemLanguage(): AppLanguage {
        val locale = Locale.getDefault()
        val tag = locale.toLanguageTag()
        if (tag.startsWith("zh")) {
            val traditional = Regex("Hant|TW|HK|MO", RegexOption.IGNORE_CASE).containsMatchIn(tag)
            return languageFor(if (traditional) "zh-Hant" else "zh-Hans") ?: english
        }
        val base = tag.split("-").first()
        languages.find { it.code == base }?.let { return it }
        if (base == "no") return languageFor("nb") ?: english
        return english
    }

    /**
     * Looks up [key] in [code], falling back to English and then to the key
     * itself so a missing entry is visible but never crashes.
     */
    fun translate(code: String, key: String): String =
        tables[code]?.get(key) ?: tables["en"]?.get(key) ?: key
}

File: web/src/lib/i18n.ts
import { useCallback } from "react";

import catalogue from "@/data/strings.json";
import { useAppStore } from "@/store/AppStore";

type StringTable = Record<string, string>;

const tables = catalogue as Record<string, StringTable>;

/** A language Calzy can present itself in. */
export interface AppLanguage {
  code: string;
  /** Name in English, so searching "Spanish" finds Español. */
  englishName: string;
  /** Endonym — what speakers call the language themselves. */
  nativeName: string;
  flag: string;
  isRTL?: boolean;
}

/** Ordered roughly by global reach, English first. Mirrors `AppLanguage.swift`. */
export const languages: AppLanguage[] = [
  { code: "en", englishName: "English", nativeName: "English", flag: "🇺🇸" },
  { code: "zh-Hans", englishName: "Chinese (Simplified)", nativeName: "简体中文", flag: "🇨🇳" },
  { code: "zh-Hant", englishName: "Chinese (Traditional)", nativeName: "繁體中文", flag: "🇹🇼" },
  { code: "es", englishName: "Spanish", nativeName: "Español", flag: "🇪🇸" },
  { code: "hi", englishName: "Hindi", nativeName: "हिन्दी", flag: "🇮🇳" },
  { code: "ar", englishName: "Arabic", nativeName: "العربية", flag: "🇸🇦", isRTL: true },
  { code: "pt", englishName: "Portuguese", nativeName: "Português", flag: "🇧🇷" },
  { code: "bn", englishName: "Bengali", nativeName: "বাংলা", flag: "🇧🇩" },
  { code: "ru", englishName: "Russian", nativeName: "Русский", flag: "🇷🇺" },
  { code: "ja", englishName: "Japanese", nativeName: "日本語", flag: "🇯🇵" },
  { code: "de", englishName: "German", nativeName: "Deutsch", flag: "🇩🇪" },
  { code: "fr", englishName: "French", nativeName: "Français", flag: "🇫🇷" },
  { code: "ko", englishName: "Korean", nativeName: "한국어", flag: "🇰🇷" },
  { code: "it", englishName: "Italian", nativeName: "Italiano", flag: "🇮🇹" },
  { code: "tr", englishName: "Turkish", nativeName: "Türkçe", flag: "🇹🇷" },
  { code: "vi", englishName: "Vietnamese", nativeName: "Tiếng Việt", flag: "🇻🇳" },
  { code: "id", englishName: "Indonesian", nativeName: "Bahasa Indonesia", flag: "🇮🇩" },
  { code: "ms", englishName: "Malay", nativeName: "Bahasa Melayu", flag: "🇲🇾" },
  { code: "th", englishName: "Thai", nativeName: "ไทย", flag: "🇹🇭" },
  { code: "pl", englishName: "Polish", nativeName: "Polski", flag: "🇵🇱" },
  { code: "nl", englishName: "Dutch", nativeName: "Nederlands", flag: "🇳🇱" },
  { code: "uk", englishName: "Ukrainian", nativeName: "Українська", flag: "🇺🇦" },
  { code: "fa", englishName: "Persian", nativeName: "فارسی", flag: "🇮🇷", isRTL: true },
  { code: "he", englishName: "Hebrew", nativeName: "עברית", flag: "🇮🇱", isRTL: true },
  { code: "el", englishName: "Greek", nativeName: "Ελληνικά", flag: "🇬🇷" },
  { code: "cs", englishName: "Czech", nativeName: "Čeština", flag: "🇨🇿" },
  { code: "ro", englishName: "Romanian", nativeName: "Română", flag: "🇷🇴" },
  { code: "hu", englishName: "Hungarian", nativeName: "Magyar", flag: "🇭🇺" },
  { code: "sv", englishName: "Swedish", nativeName: "Svenska", flag: "🇸🇪" },
  { code: "nb", englishName: "Norwegian", nativeName: "Norsk", flag: "🇳🇴" },
  { code: "da", englishName: "Danish", nativeName: "Dansk", flag: "🇩🇰" },
  { code: "fi", englishName: "Finnish", nativeName: "Suomi", flag: "🇫🇮" },
];

const english: AppLanguage = languages[0];

export function languageFor(code: string | undefined): AppLanguage | undefined {
  if (code === undefined) return undefined;
  return languages.find((language) => language.code === code);
}

/**
 * Best match for the browser's preferred languages, falling back to English.
 *
 * Chinese needs script-level matching (`zh-Hans` vs `zh-Hant`); everything else
 * matches on the base language subtag.
 */
export function browserLanguage(): AppLanguage {
  const preferred = typeof navigator === "undefined" ? [] : navigator.languages ?? [navigator.language];
  for (const tag of preferred) {
    if (tag.startsWith("zh")) {
      const isTraditional = /Hant|TW|HK|MO/i.test(tag);
      return languageFor(isTraditional ? "zh-Hant" : "zh-Hans") ?? english;
    }
    const base = tag.split("-")[0];
    const match = languages.find((language) => language.code === base);
    if (match) return match;
    if (base === "no") return languageFor("nb") ?? english;
  }
  return english;
}

/**
 * Looks up `key` in `code`, falling back to English and then to the key itself
 * so a missing entry is visible but never throws.
 */
export function translate(code: string, key: string): string {
  return tables[code]?.[key] ?? tables.en?.[key] ?? key;
}

/** Active language plus a `t()` bound to it. Re-renders when the choice changes. */
export function useLanguage(): { language: AppLanguage; t: (key: string) => string } {
  const store = useAppStore();
  const language = languageFor(store.profile.languageCode) ?? browserLanguage();
  const code = language.code;
  const t = useCallback((key: string) => translate(code, key), [code]);
  return { language, t };
}

/** Shorthand for components that only need the translator. */
export function useT(): (key: string) => string {
  return useLanguage().t;
}
